tools/publish-aos.js

```diff
@@ -1,54 +1,137 @@
+  const matches = changelog.match(sourceCodeTxIdRegex);
+
+  const sourceCodeTxIds = new Set(matches ?? []);
+
+  return [...sourceCodeTxIds];
+}
 async function main() {
+  // get the last published source code from the changelogs folder (can check the date on it, check for multiple dates) and
+  // fetch the code from arweave, then compare it with the current code
+
+  const publishedSourceCodeIds = getPublishedSourceCodeIds();
+  const lastPublishedSourceCodeId = await arweave.api
+    .post('/graphql', {
+      query: `{
+      transactions(
+        first: 1
+        sort: HEIGHT_DESC
+        ids: [${publishedSourceCodeIds.map((id) => `"${id}"`).join(',')}]
+      ) {
+        edges {
+          node {
+            id
+            tags {
+              name
+              value
+            }
+            block {
+              timestamp
+            }
+          }
+        }
+      }
+    }`,
+    })
+    .then((res) => {
+      // TODO: add pagination once we have more than 100 versions
+      if (res.data.data.transactions.edges.length > 99)
+        throw new Error('Too many transactions found, implement pagination');
+      return res.data.data.transactions.edges[0].node.tags.find(
+        (tag) => tag.name === 'Original-Tx-Id',
+      )?.value;
+    });
+
+  const lastPublishedSourceCode = await arweave.transactions
+    .getData(lastPublishedSourceCodeId, { decode: true, string: true })
+    .catch((error) => {
+      console.error(
+        'Error fetching last published source code, unable to compare with current code. \n',
+        error,
+      );
+    });
+
   const bundledLua = fs.readFileSync(
     path.join(__dirname, '../dist/aos-bundled.lua'),
     'utf-8',
   );
-  const wallet = fs.readFileSync(path.join(__dirname, 'key.json'), 'utf-8');
+
+  if (lastPublishedSourceCode === bundledLua) {
+    console.log('No changes in source code detected, skipping publishing');
+    process.exit(0); // Exit with code 0 to indicate success
+  }
+  let wallet = process.env.WALLET;
+  if (!wallet) {
+    try {
+      wallet = fs.readFileSync(path.join(__dirname, 'key.json'), 'utf-8');
+    } catch (error) {
+      console.error(
+        'No key found, please set ARWEAVE_PUBLISHING_KEY environment variable or provide a key.json file',
+      );
+      return;
+    }
+  }
   const jwk = JSON.parse(wallet);
-  const address = await arweave.wallets.jwkToAddress(jwk);
+  const signer = new ArweaveSigner(jwk);
+  const turbo = TurboFactory.authenticated({ signer });
 
-  const tx1 = await arweave.createTransaction({ data: bundledLua }, jwk);
-  tx1.addTag('App-Name', 'aos-LUA');
-  tx1.addTag('App-Version', '0.0.1');
-  tx1.addTag('Content-Type', 'text/x-lua');
-  tx1.addTag('Author', 'Permanent Data Solutions');
-  await arweave.transactions.sign(tx1, jwk);
-  await arweave.transactions.post(tx1);
+  const publishingTags = Object.entries({
+    'App-Name': 'aos-LUA',
+    'App-Version': '8',
```

Comment:
pull this from a constants file version.js similar to ar-io-node repo

Reply:
you ONLY need to do the first part 

> pull this from a constants file version.js similar to ar-io-node repo

the later is comment on how we'll eventually connect them and can be done separately